nums/core/array/application.py

```diff
@@ -332,28 +331,37 @@ def diag(self, X: BlockArray) -> BlockArray:
                 if np.all(np.diff(grid_entry) == 0):
                     # This is a diagonal block.
-                    rarr.blocks[grid_entry].oid = self.cm.diag(X.blocks[grid_entry[0]].oid,
+                    rarr.blocks[grid_entry].oid = self.cm.diag(X.blocks[grid_entry[0]].oid, 0,
                                                                syskwargs=syskwargs)
                 else:
                     rarr.blocks[grid_entry].oid = self.cm.new_block("zeros",
                                                                     grid_entry,
                                                                     grid_meta,
                                                                     syskwargs=syskwargs)
         elif len(X.shape) == 2:
-            assert X.shape[0] == X.shape[1], "X must be a square array."
-            assert X.block_shape[0] == X.block_shape[1], "block_shape must be square."
-            shape = X.shape[0],
-            block_shape = X.block_shape[0],
-            grid = ArrayGrid(shape, block_shape, X.dtype.__name__)
-            rarr = BlockArray(grid, self.cm)
-            for grid_entry in X.grid.get_entry_iterator():
-                out_grid_entry = grid_entry[:1]
-                out_grid_shape = grid.grid_shape[:1]
-                syskwargs = {"grid_entry": out_grid_entry, "grid_shape": out_grid_shape}
-                if np.all(np.diff(grid_entry) == 0):
-                    # This is a diagonal block.
-                    rarr.blocks[out_grid_entry].oid = self.cm.diag(X.blocks[grid_entry].oid,
-                                                                   syskwargs=syskwargs)
+            out_shape = min(X.shape),
+            out_block_shape = min(X.block_shape),
+            # Obtain the block indices which contain the diagonal of the matrix.
+            diag_meta = array_utils.find_output_blocks(X.blocks, out_shape[0])
+
+            all_result_blocks = []
+            out_grid_shape, count = (len(diag_meta),), 0
+            # Obtain the diagonals.
+            for block_indices, offset, total_elements in diag_meta:
+                syskwargs = {"grid_entry": (count,) ,"grid_shape": out_grid_shape}
+                result_block_shape = total_elements,
+                block_grid = ArrayGrid(result_block_shape, result_block_shape,
+                                        X.blocks[block_indices].dtype.__name__)
+                block = BlockArray(block_grid, self.cm)
+                block.blocks[0].oid = self.cm.diag(X.blocks[block_indices].oid,
+                                                 offset, syskwargs=syskwargs)
+                all_result_blocks.append(block)
+                count += 1
+            if len(all_result_blocks) > 1:
+                # If there are multiple blocks, concatenate them.
+                return self.concatenate(all_result_blocks, axis=0,
+                                        axis_block_size=out_block_shape[0])
+            return block
```

Comment:
Instead of returning `block`, can you return `all_result_blocks[0]`? While it's true in Python, in general it's not good practice to assume variables defined in loops exist outside of the scope of the loop.

Reply:
Yeah I was thinking about the same. But, I have changed it now.